distributed/shuffle/_merge.py

```diff
@@ -0,0 +1,447 @@
+            except ValueError:
+                continue
+            if _name != self.name:
+                continue
+            parts.add(_part)
+        return parts
+
+    def get_output_keys(self):
+        return {(self.name, part) for part in self.parts_out}
+
+    def __repr__(self):
+        return f"HashJoin<name='{self.name}', npartitions={self.npartitions}>"
+
+    def is_materialized(self):
+        return hasattr(self, "_cached_dict")
+
+    def __getitem__(self, key):
+        return self._dict[key]
+
+    def __iter__(self):
+        return iter(self._dict)
+
+    def __len__(self):
+        return len(self._dict)
+
+    @property
+    def _dict(self):
+        """Materialize full dict representation"""
+        if hasattr(self, "_cached_dict"):
+            return self._cached_dict
+        else:
+            dsk = self._construct_graph()
+            self._cached_dict = dsk
+        return self._cached_dict
+
+    def _cull(self, parts_out: Sequence[str]):
+        return HashJoinP2PLayer(
+            name=self.name,
+            name_input_left=self.name_input_left,
+            left_on=self.left_on,
+            name_input_right=self.name_input_right,
+            right_on=self.right_on,
+            how=self.how,
+            npartitions=self.npartitions,
+            suffixes=self.suffixes,
+            indicator=self.indicator,
+            meta_output=self.meta_output,
+            parts_out=parts_out,
+            left_index=self.left_index,
+            right_index=self.right_index,
+            annotations=self.annotations,
+            n_partitions_left=self.n_partitions_left,
+            n_partitions_right=self.n_partitions_right,
+        )
+
+    def cull(self, keys: Iterable[str], all_keys: Any) -> tuple[HashJoinP2PLayer, dict]:
+        """Cull a SimpleShuffleLayer HighLevelGraph layer.
+
+        The underlying graph will only include the necessary
+        tasks to produce the keys (indices) included in `parts_out`.
+        Therefore, "culling" the layer only requires us to reset this
+        parameter.
+        """
+        parts_out = self._keys_to_parts(keys)
+        culled_deps = self._cull_dependencies(keys, parts_out=parts_out)
+        if parts_out != set(self.parts_out):
+            culled_layer = self._cull(parts_out)
+            return culled_layer, culled_deps
+        else:
+            return self, culled_deps
+
+    def _construct_graph(self) -> dict[tuple | str, tuple]:
+        args = (
+            self.left_on,
+            self.how,
+            self.npartitions,
+            self.n_partitions_left,
+            self.n_partitions_right,
+            self.parts_out,
+            self.suffixes,
+            self.indicator,
+        )
+        token_left = tokenize(self.name_input_left, *args)
+        token_right = tokenize(self.name_input_right, *args)
+        dsk: dict[tuple | str, tuple] = {}
+        # FIXME: This is a problem. The barrier key is parsed to infer the
```

Comment:
What issues does this cause apart from being a clumsy API?

Reply:
I was concerned that the transitions hooks (or anything that relies on parsing the data) would not be able to restart both shuffles if necessary, e.g. when a worker leaves the set.
Not sure if this is actually relevant.